internal/envoy/route.go

```diff
@@ -194,6 +194,44 @@ func UpgradeHTTPS() *envoy_api_v2_route.Route_Redirect {
 	}
 }
 
+// HeaderValueList creates a list of Envoy HeaderValueOptions from the provided map.
+func HeaderValueList(hvm map[string]string, app bool) (hvs []*envoy_api_v2_core.HeaderValueOption) {
+	for key, value := range hvm {
+		hvs = append(hvs, &envoy_api_v2_core.HeaderValueOption{
+			Header: &envoy_api_v2_core.HeaderValue{
+				Key:   key,
+				Value: value,
+			},
+			Append: &wrappers.BoolValue{
+				Value: app,
+			},
+		})
+	}
+	sort.Slice(hvs, func(i, j int) bool {
+		return hvs[i].Header.Key < hvs[j].Header.Key
+	})
+	return
+}
+
+// singleSimpleCluster determines whether we can use a RouteAction_Cluster
+// or must use a RouteAction_WeighedCluster to encode additional routing data.
+func singleSimpleCluster(clusters []*dag.Cluster) bool {
```

Comment:
Why do you need to use a weighted cluster if any of the header additions are used?

Reply:
The non-weighted version isn't expressive enough because it just takes a cluster name:
```go
type RouteAction_Cluster struct {
	Cluster string `protobuf:"bytes,1,opt,name=cluster,proto3,oneof"`
}
```

However, the `RouteAction_WeightedClusters` takes a list of clusters with additional stuff to attach.  I'd guess this is creep in the Envoy API that led to the container type being a misnomer, but it'd take some archaeology to confirm.

Reply:
That's unfortunate that it only applied to Weighted Clusters. We had thoughts around removing weighted clusters in the future to solve different problems with healthchecks.